File: version_3/PSO.R
# ==============================================================================
# PSO.R — PSO con Restricciones de Cardinalidad
# ==============================================================================

library(cluster)
library(pso)
library(proxy)
library(dplyr)

if (!requireNamespace("aricode", quietly = TRUE)) install.packages("aricode")
library(aricode)

# -----------------------------
# Registro interno de hiperparámetros (solo en memoria)
# Consultar: .pso_hyperparams$global / .pso_hyperparams$runs[["nombre"]]
# -----------------------------
.pso_hyperparams <- list(
  global = list(
    algorithm          = "PSO",
    distance_method    = "cosine",
    pso_maxit          = 20L,
    pso_swarm_size     = 40L,
    par_lower          = 0,
    par_upper          = 1,
    objective          = "intra_cluster_cosine_distance",
    assignment_method  = "order_based_partition"
  ),
  runs = list()
)

compute_metrics <- function(y_true, y_predict, X) {
  y_true_int    <- as.integer(as.factor(y_true))
  y_predict_int <- as.integer(y_predict)
  ARI_val <- tryCatch(aricode::ARI(y_true_int, y_predict_int),  error = function(e) NA_real_)
  AMI_val <- tryCatch(aricode::AMI(y_true_int, y_predict_int),  error = function(e) NA_real_)
  NMI_val <- tryCatch(aricode::NMI(y_true_int, y_predict_int),  error = function(e) NA_real_)
  Sil_val <- tryCatch({
    if (length(unique(y_predict_int)) < 2) return(NA_real_)
    d   <- proxy::dist(as.matrix(X), method = "cosine")
    sil <- cluster::silhouette(y_predict_int, d)
    mean(sil[, "sil_width"], na.rm = TRUE)
  }, error = function(e) NA_real_)
  list(ARI = ARI_val, AMI = AMI_val, NMI = NMI_val, Silhouette_mean = Sil_val)
}

# Salida
dir.create("predictions", showWarnings = FALSE, recursive = FALSE)
.pred_pso_csv <- "predictions/pred_PSO.csv"

# -----------------------------
# Helpers robustos
# -----------------------------
make_numeric_X <- function(X) {
  X <- as.data.frame(X)
  
  X_num <- lapply(X, function(col) {
    if (is.numeric(col) || is.integer(col)) return(as.numeric(col))
    if (is.logical(col)) return(as.numeric(col))
    if (is.factor(col))  return(as.numeric(col))
    if (is.character(col)) return(as.numeric(as.factor(col)))
    return(NULL)
  })
  
  keep <- !vapply(X_num, is.null, logical(1))
  X_num <- X_num[keep]
  if (length(X_num) == 0) return(NULL)
  
  X_num <- as.data.frame(X_num)
  
  X_num <- X_num[, vapply(X_num, function(z) length(unique(z)) > 1, logical(1)), drop = FALSE]
  if (ncol(X_num) == 0) return(NULL)
  
  X_num
}

prepare_data <- function(dataset) {
  dataset <- as.data.frame(dataset)
  if (ncol(dataset) < 2) return(NULL)
  
  if ("class" %in% colnames(dataset)) {
    y <- dataset$class
    X <- dataset[, setdiff(colnames(dataset), "class"), drop = FALSE]
  } else {
    y <- dataset[[ncol(dataset)]]
    X <- dataset[, -ncol(dataset), drop = FALSE]
  }
  
  X_num <- make_numeric_X(X)
  if (is.null(X_num)) return(NULL)
  
  list(X = X_num, y = as.factor(y))
}

# -----------------------------
# PSO
# OPT: pre-calcular cumsum de cluster_sizes para asignación sin loop
# OPT: cost_function usa la cumsum pre-calculada
# -----------------------------
run_PSO <- function(X, target_cardinality) {
  
  set.seed(123)
  Xmat <- as.matrix(X)
  
  if (!all(is.finite(Xmat))) stop("X contiene NA/Inf")
  if (!all(vapply(as.data.frame(Xmat), is.numeric, logical(1)))) stop("X debe ser numérico")
  
  n <- nrow(Xmat)
  k <- length(target_cardinality)
  
  if (k < 2) stop("k < 2")
  if (sum(target_cardinality) != n) stop("sum(target_cardinality) != n (inconsistente)")
  
  D <- as.matrix(proxy::dist(Xmat, method = "cosine"))
  
  # OPT: pre-calcular los índices de corte para evitar loop en cost_function
  cs_sizes <- as.integer(target_cardinality)
  cum_sizes <- cumsum(cs_sizes)
  start_positions <- c(1L, cum_sizes[-k] + 1L)
  
  cost_function <- function(par, dist_matrix, cluster_sizes, k) {
    n <- length(par)
    cluster_assignment <- integer(n)
    
    ord <- order(par)
    # OPT: usar start/end pre-calculados
    for (i in 1:k) {
      cluster_assignment[ord[start_positions[i]:cum_sizes[i]]] <- i
    }
    
    total_distance <- 0
    for (i in 1:k) {
      ci <- which(cluster_assignment == i)
      if (length(ci) > 1) {
        cd <- dist_matrix[ci, ci, drop = FALSE]
        total_distance <- total_distance + sum(cd[lower.tri(cd, diag = FALSE)])
      }
    }
    total_distance
  }
  
  pso_result <- psoptim(
    par = runif(n),
    fn = cost_function,
    dist_matrix = D,
    cluster_sizes = cs_sizes,
    k = k,
    lower = 0,
    upper = 1,
    control = list(maxit = 20, s = 40)
  )
  
  label_pred <- integer(n)
  ord <- order(pso_result$par)
  # OPT: usar start/end pre-calculados
  for (i in seq_len(k)) {
    label_pred[ord[start_positions[i]:cum_sizes[i]]] <- i
  }
  
  list(y_predict = label_pred)
}

# -----------------------------
# Runner: devuelve fila
# -----------------------------
run_clustering_row <- function(dataset, target_cardinality, dataset_name) {
  
  set.seed(123)
  data <- prepare_data(dataset)
  if (is.null(data)) return(NULL)
  
  X <- data$X
  y <- data$y
  
  if (is.null(target_cardinality) || all(is.na(target_cardinality))) return(NULL)
  target_cardinality <- as.integer(target_cardinality)
  
  if (sum(target_cardinality) != nrow(X)) return(NULL)
  if (length(target_cardinality) < 2) return(NULL)
  
  start_total <- proc.time()
  
  results <- tryCatch(run_PSO(X, target_cardinality), error = function(e) NULL)
  
  if (is.null(results) || is.null(results$y_predict)) return(NULL)
  
  y_predict <- results$y_predict
  y_int <- as.integer(y)
  
  total_time <- (proc.time() - start_total)[3]
  
  metrics <- compute_metrics(y, y_predict, X)
  
  # Guardar hiperparámetros de esta ejecución (solo en memoria)
  .pso_hyperparams$runs[[dataset_name]] <<- list(
    k                = length(target_cardinality),
    n                = length(y_predict),
    n_features       = ncol(as.matrix(X)),
    target_sizes     = target_cardinality,
    pso_maxit        = 20L,
    pso_swarm_size   = 40L,
    par_lower        = 0,
    par_upper        = 1,
    distance_method  = "cosine",
    assignment_method = "order_based_partition"
  )
  
  data.frame(
    name = dataset_name,
    n = length(y_predict),
    k = length(unique(y_predict)),
    y_predict = paste(as.integer(y_predict), collapse = " "),
    y_true = paste(as.integer(y_int), collapse = " "),
    target_cardinality = paste(as.integer(target_cardinality), collapse = " "),
    cardinality_pred   = paste(as.integer(table(factor(y_predict,
                                                       levels = 1:length(target_cardinality)))), collapse = " "),
    Execution_Time = as.numeric(total_time),
    ARI = metrics$ARI,
    AMI = metrics$AMI,
    NMI = metrics$NMI,
    Silhouette_mean = metrics$Silhouette_mean,
    stringsAsFactors = FALSE
  )
}

# -----------------------------
# Loop principal
# OPT: pre-asignar lista, seq_len, vapply
# -----------------------------
results_list <- vector("list", nrow(odatasets_unique))

for (i in seq_len(nrow(odatasets_unique))) {
  cat("\n\n--- Executing PSO for dataset at position:", i, "---\n")
  
  tryCatch({
    dataset <- odatasets_unique$dataset[[i]]
    dataset_name <- odatasets_unique$name[[i]]
    target_cardinality <- odatasets_unique$class_distribution_vector[[i]]
    
    if (is.null(dataset) || is.null(target_cardinality) || all(is.na(target_cardinality))) {
      cat("Dataset o cardinalidad inválida. Skipping.\n")
      next
    }
    
    row_result <- run_clustering_row(dataset, target_cardinality, dataset_name)
    
    if (!is.null(row_result)) {
      results_list[[i]] <- row_result
      cat("Time:", row_result$Execution_Time, "seconds\n")
    } else {
      cat("Dataset skipped.\n")
    }
    
  }, error = function(e) {
    cat("Error:", e$message, "\n")
  })
}

# -----------------------------
# Escritura final
# -----------------------------
results_clean <- results_list[!vapply(results_list, is.null, logical(1))]

if (length(results_clean) > 0) {
  final_df <- do.call(rbind, results_clean)
  write.table(final_df, .pred_pso_csv, sep = ",", row.names = FALSE, col.names = TRUE)
  cat("\nPSO finalizado. Archivo guardado exitosamente en:", .pred_pso_csv, "\n")
} else {
  cat("\nPSO finalizado sin resultados válidos. No se creó el CSV.\n")
}

File: clustering-python/algorithms/pso.py
"""
algorithms/pso.py
=================
PSO (Particle Swarm Optimization) con restricciones de cardinalidad.

Estructura del algoritmo:
  - Cada partícula es un vector de N reales en [0, 1].
  - Codificación order-based: ordenamos los valores y partimos en k segmentos
    de tamaños target[1], target[2], ..., target[k]. Garantiza cardinalidad
    exacta sin ajustes post-hoc (a diferencia de BAT y ACO).
  - Función de costo: suma de distancias intra-cluster (coseno).
  - 40 partículas × 20 iteraciones.
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable, Optional, cast

import numpy as np
import pandas as pd
from scipy.spatial.distance import pdist, squareform

from ._common import (
    compute_metrics,
    get_predictions_dir,
    prepare_data,
    tabulate_clusters,
    to_int_list,
)

logger = logging.getLogger(__name__)
if not logger.handlers:
    logging.basicConfig(level=logging.INFO, format="%(message)s")


# ============================================================================
# Hiperparámetros y rutas de salida
# ============================================================================

@dataclass
class PsoHyperparams:
    """Hiperparámetros globales del algoritmo PSO."""
    algorithm:         str   = "PSO"
    distance_method:   str   = "cosine"
    pso_maxit:         int   = 20
    pso_swarm_size:    int   = 40
    par_lower:         float = 0.0
    par_upper:         float = 1.0
    objective:         str   = "intra_cluster_cosine_distance"
    assignment_method: str   = "order_based_partition"
    master_seed:       int   = 123


PSO_HYPERPARAMS: dict[str, Any] = {
    "global": PsoHyperparams(),
    "runs":   {},
}

# La ruta de salida se obtiene en runtime via get_predictions_dir()
# de _common, que lee la env var CLUSTERING_MODEL seteada por testing.py.


# ============================================================================
# Codificación order-based: vector real -> asignación de cluster
# ============================================================================

def _decode_par_to_clusters(
    par:             np.ndarray,
    k:               int,
    start_positions: np.ndarray,
    cum_sizes:       np.ndarray,
) -> np.ndarray:
    """
    Decodifica un vector real en [0,1] a una asignación de cluster (1-based).

    Ordena `par` ascendentemente y parte el vector ordenado en k segmentos
    de tamaños target[1], target[2], ..., target[k]:
      - posiciones [0, target[0])         -> cluster 1
      - posiciones [target[0], target[0]+target[1]) -> cluster 2
      - ...
    """
    cluster_assignment = np.zeros(len(par), dtype=int)
    ord_idx = np.argsort(par)
    for i in range(k):
        cluster_assignment[ord_idx[start_positions[i]:cum_sizes[i]]] = i + 1
    return cluster_assignment


# ============================================================================
# Función de costo: suma de distancias intra-cluster
# ============================================================================

def _cost_fn(
    par:             np.ndarray,
    dist_matrix:     np.ndarray,
    k:               int,
    start_positions: np.ndarray,
    cum_sizes:       np.ndarray,
) -> float:
    """
    Suma de distancias intra-cluster.

    Trick: como la matriz de distancias es simétrica con diagonal 0,
    sum(lower_tri) = sum(matriz) / 2. Esto es ~2x más rápido que np.tril.
    """
    cluster_assignment = _decode_par_to_clusters(par, k, start_positions, cum_sizes)
    total = 0.0
    for i in range(1, k + 1):
        ci = np.where(cluster_assignment == i)[0]
        if len(ci) > 1:
            cd = dist_matrix[np.ix_(ci, ci)]
            # cd simétrica con diagonal=0  =>  lower_tri.sum() = cd.sum() / 2
            total += float(cd.sum() * 0.5)
    return total


# ============================================================================
# PSO global-best (defaults SPSO 2007 de R psoptim)
# ============================================================================

def _pso_optimize(
    cost_fn:    Callable[[np.ndarray], float],
    n_dim:      int,
    swarm_size: int,
    max_iter:   int,
    lower:      float,
    upper:      float,
    rng:        np.random.Generator,
) -> tuple[np.ndarray, float]:
    """
    PSO global-best básico con defaults de R psoptim (SPSO 2007).

    Devuelve (mejor_posicion, mejor_score). Score se MINIMIZA (igual que R).
    """
    # Hiperparámetros estándar SPSO 2007 (defaults de R psoptim)
    w  = 1.0 / (2.0 * np.log(2.0))   # ≈ 0.7213
    c1 = 0.5 + np.log(2.0)            # ≈ 1.1931
    c2 = 0.5 + np.log(2.0)            # ≈ 1.1931

    # Inicializar posiciones uniformes en [lower, upper]
    positions:  np.ndarray = rng.uniform(lower, upper, size=(swarm_size, n_dim))
    velocities: np.ndarray = np.zeros((swarm_size, n_dim), dtype=float)

    # Evaluar posiciones iniciales
    pbest_scores: np.ndarray = np.array(
        [cost_fn(positions[i]) for i in range(swarm_size)],
        dtype=float,
    )
    pbest: np.ndarray = positions.copy()

    g_idx       = int(np.argmin(pbest_scores))
    gbest:      np.ndarray = pbest[g_idx].copy()
    gbest_score = float(pbest_scores[g_idx])

    for _ in range(max_iter):
        # Actualización de velocidades VECTORIZADA sobre todo el enjambre
        r1 = rng.random((swarm_size, n_dim))
        r2 = rng.random((swarm_size, n_dim))
        velocities = (
            w * velocities
            + c1 * r1 * (pbest - positions)
            + c2 * r2 * (gbest[np.newaxis, :] - positions)
        )
        positions = positions + velocities
        positions = np.clip(positions, lower, upper)

        # Evaluación: secuencial (cost_fn no se vectoriza fácilmente porque
        # cada partícula tiene un order distinto)
        for i in range(swarm_size):
            score = cost_fn(positions[i])
            if score < pbest_scores[i]:
                pbest[i]        = positions[i].copy()
                pbest_scores[i] = score
                if score < gbest_score:
                    gbest_score = score
                    gbest       = positions[i].copy()

    return gbest, gbest_score


# ============================================================================
# Algoritmo PSO completo
# ============================================================================

def run_pso_algorithm(
    x_mat: np.ndarray,
    target_cardinality: np.ndarray,
    swarm_size:  int   = 40,
    max_iter:    int   = 20,
    par_lower:   float = 0.0,
    par_upper:   float = 1.0,
    master_seed: int   = 123,
) -> dict[str, Any]:
    """
    Ejecuta PSO sobre x_mat. Devuelve dict con y_predict, best_score y full_dist.
    """
    rng    = np.random.default_rng(master_seed)
    x_mat  = np.asarray(x_mat, dtype=float)
    n      = int(x_mat.shape[0])
    target = np.asarray(target_cardinality, dtype=int)
    k      = int(len(target))

    # Validaciones equivalentes al R original
    if not np.all(np.isfinite(x_mat)):
        raise ValueError("X contiene NA/Inf")
    if k < 2:
        raise ValueError("k < 2")
    if int(target.sum()) != n:
        raise ValueError(
            f"sum(target_cardinality)={int(target.sum())} != n={n} (inconsistente)"
        )

    # ----- Distancia coseno N×N (reutilizada para silhouette luego) -----
    # Bug del stub de scipy: pdist sí acepta strings como `metric`.
    condensed = cast(np.ndarray, pdist(x_mat, metric="cosine"))  # pyright: ignore[reportArgumentType, reportCallIssue]
    full_dist = squareform(condensed)

    # ----- Pre-cálculo de cortes (OPT del R original, mantenido) -----
    cum_sizes       = np.cumsum(target)
    start_positions = np.concatenate([[0], cum_sizes[:-1]])

    # Closure que captura D, target, start_positions, cum_sizes
    def cost_fn(par: np.ndarray) -> float:
        return _cost_fn(par, full_dist, k, start_positions, cum_sizes)

    # ----- Optimización PSO -----
    best_par, best_score = _pso_optimize(
        cost_fn    = cost_fn,
        n_dim      = n,
        swarm_size = swarm_size,
        max_iter   = max_iter,
        lower      = par_lower,
        upper      = par_upper,
        rng        = rng,
    )

    # ----- Decodificar la mejor partícula -----
    y_pred = _decode_par_to_clusters(best_par, k, start_positions, cum_sizes)

    return {
        "y_predict":  y_pred,
        "best_score": best_score,
        "full_dist":  full_dist,
    }


# ============================================================================
# Runner por dataset
# ============================================================================

def run_clustering_row(
    dataset: pd.DataFrame,
    target_cardinality: list[int],
    dataset_name: str,
) -> Optional[dict[str, Any]]:
    """Ejecuta PSO sobre un dataset y devuelve la fila de resultados."""
    data = prepare_data(dataset)
    if data is None:
        return None

    x_df:     pd.DataFrame    = data["X"]
    y_factor: pd.Categorical = data["y"]

    target = np.asarray(target_cardinality, dtype=int)
    x_mat  = x_df.to_numpy(dtype=float)
    n      = int(x_mat.shape[0])
    k      = int(len(target))

    if k < 2 or int(target.sum()) != n:
        return None

    start_total = time.perf_counter()

    try:
        result = run_pso_algorithm(x_mat, target)
    except Exception as e:  # pylint: disable=broad-exception-caught
        logger.warning("Error en run_pso_algorithm para %s: %s", dataset_name, e)
        return None

    y_pred:    np.ndarray = result["y_predict"]
    full_dist: np.ndarray = result["full_dist"]
    y_int                  = (y_factor.codes + 1).astype(int)
    total_time             = time.perf_counter() - start_total

    metrics = compute_metrics(y_int, y_pred, x_mat, dist_matrix=full_dist)

    PSO_HYPERPARAMS["runs"][dataset_name] = {
        "k":                 k,
        "n":                 n,
        "n_features":        x_mat.shape[1],
        "target_sizes":      target.tolist(),
        "pso_maxit":         20,
        "pso_swarm_size":    40,
        "par_lower":         0.0,
        "par_upper":         1.0,
        "distance_method":   "cosine",
        "assignment_method": "order_based_partition",
    }

    cardinality_pred = tabulate_clusters(y_pred, k)

    return {
        "name":               dataset_name,
        "n":                  len(y_pred),
        "k":                  int(len(np.unique(y_pred))),
        "y_predict":          " ".join(str(int(c)) for c in y_pred),
        "y_true":             " ".join(str(int(c)) for c in y_int),
        "target_cardinality": " ".join(str(int(c)) for c in target),
        "cardinality_pred":   " ".join(str(int(c)) for c in cardinality_pred),
        "Execution_Time":     float(total_time),
        "ARI":                metrics["ARI"],
        "AMI":                metrics["AMI"],
        "NMI":                metrics["NMI"],
        "Silhouette_mean":    metrics["Silhouette_mean"],
    }


# ============================================================================
# Punto de entrada para testing.py
# ============================================================================

def run(odatasets_unique: pd.DataFrame) -> None:
    """Loop principal equivalente al de PSO.R."""
    results: list[dict[str, Any]] = []

    for i in range(len(odatasets_unique)):
        print(
            f"\n--- Executing PSO for dataset at position: {i + 1} ---",
            flush=True,
        )
        try:
            dataset      = odatasets_unique["dataset"].iloc[i]
            dataset_name = odatasets_unique["name"].iloc[i]
            target_raw   = odatasets_unique["class_distribution_vector"].iloc[i]

            target = to_int_list(target_raw)
            if dataset is None or target is None or len(target) < 2:
                print("Dataset o cardinalidad inválida. Skipping.")
                continue

            row = run_clustering_row(dataset, target, dataset_name)

            if row is not None:
                results.append(row)
                print(f"Time: {row['Execution_Time']:.4f} seconds")
            else:
                print("Dataset skipped.")

        except Exception as e:  # pylint: disable=broad-exception-caught
            print(f"Error: {e}")

    if results:
        df_out = pd.DataFrame(results)
        out_path = get_predictions_dir() / "pred_PSO.csv"
        df_out.to_csv(out_path, index=False)
        print(f"\nPSO finalizado. Archivo guardado en: {out_path}")
    else:
        print("\nPSO finalizado sin resultados válidos. No se generó CSV.")
